tests/cmdline/commands/test_storage.py

```diff
@@ -0,0 +1,118 @@
+    from aiida.manage.manager import get_manager
+    profile = get_manager().get_profile()
+
+    result = run_cli_command(cmd_storage.backend_migrate, user_input='MIGRATE NOW')
+
+    assert 'warning' in result.output.casefold()
+    assert profile.name in result.output
+    assert 'migrate now' in result.output.casefold()
+    assert 'migration completed' in result.output.casefold()
+
+
+def tests_storage_migrate_running_daemon(run_cli_command, monkeypatch):
+    """Test that ``verdi storage migrate`` raises if the daemon is running."""
+    from aiida.engine.daemon.client import DaemonClient
+
+    def return_true():
+        return True
+
+    with monkeypatch.context() as mock_context:
+        mock_context.setattr(DaemonClient, 'is_daemon_running', return_true)
+        result = run_cli_command(cmd_storage.backend_migrate, raises=True)
+
+    assert 'daemon' in result.output.casefold()
+    assert 'running' in result.output.casefold()
+
+
+def tests_storage_migrate_cancel_prompt(run_cli_command, monkeypatch):
+    """Test that ``verdi storage migrate`` detects the cancelling of the interactive prompt."""
+    import click
+
+    def raise_click_abort(*args, **kwargs):
+        raise click.Abort()
+
+    with monkeypatch.context() as mock_context:
+        mock_context.setattr(click, 'prompt', raise_click_abort)
+        result = run_cli_command(cmd_storage.backend_migrate, raises=True)
+
+    assert 'aborted' in result.output.casefold()
+
+
+@pytest.mark.parametrize('raise_type', [
+    exceptions.ConfigurationError,
+    exceptions.DatabaseMigrationError,
+])
+@pytest.mark.parametrize(
+    'call_kwargs', [
+        {
+            'raises': True,
+            'user_input': 'MIGRATE NOW'
+        },
+        {
+            'raises': True,
+            'options': ['--force']
+        },
+    ]
+)
+def tests_storage_migrate_raises(run_cli_command, raise_type, call_kwargs, monkeypatch):
+    """Test that ``verdi storage migrate`` detects errors while migrating.
+
+    Note: I can't monkeypatch the result of `manager.get_backend` because the `backend_migrate`
+    function actually calls `manager._load_backend` that will force the re-instantiation of
+    the backend member of the manager, thus overwritting the patch. Instead I monkeypatched
+    the `backend_migrate` directly so that it would return an object whose `migrate` method
+    raises.
+    """
+    from aiida.manage.manager import get_manager
+    manager = get_manager()
+
+    def return_raiser(*args, **kwargs):
+
+        class BackendRaiser():
+
+            def migrate(self):  # pylint: disable=no-self-use
+                raise raise_type('passed error message')
+
+        return BackendRaiser()
+
+    with monkeypatch.context() as mock_context:
+        mock_context.setattr(manager, '_load_backend', return_raiser)
+        result = run_cli_command(cmd_storage.backend_migrate, **call_kwargs)
+
+    assert 'passed error message' in result.output
```

Comment:
Not sure what you are testing here. You are forcing it to raise an exception with a specific string and then checking that that string is in the output of the CLI command. But that would be a given, wouldn't it? Probably you want to check that the exception is _caught_ and a warning message is shown instead of letting the exception bubble up and show the stack trace. In that case you have to make sure that the output does _not_ contain the stack trace or you have to make your positive check more specific and not just look for just the exception message.

Reply:
Well, actually that is a very good question. What is the point of catching [this](https://github.com/aiidateam/aiida-core/pull/5228/files#diff-f3d13a2ae78add210c0cfb8765a16c3de937fd3fe0cac54a35a3d23928e9b953R40-R45)?

``` python
        try:
            backend.migrate()
        except (exceptions.ConfigurationError, exceptions.DatabaseMigrationError) as exception:
            echo.echo_critical(str(exception))
```

It is not adding any contextual information to help the user identify the problem, it is not doing anything afterwards. What is the point of catching to basically re-raise? Just to redirect to the `CMDLINE_LOGGER` in case is not set to stdout/stderr?

Reply:
The point is that these are known failure modes and if they occur, we want to show the error message, but not just have the CLI command except. I catch them explicitly (and not just broad `Exception`) because any other exceptions are not expected and constitute a bug in our code.

Reply:
Yes, perfect.